src/rulesets/LinearAlgebra/structured.jl

```diff
@@ -19,6 +19,13 @@ function rrule(::typeof(diag), A::AbstractMatrix)
     return diag(A), diag_pullback
 end
 
+function rrule(::typeof(*), D::Diagonal{<:Real}, V::AbstractVector{<:Real})
+    function times_pullback(Ȳ)
```

Comment:
I'm not very familiar with this package, is there a reason why this sub-function is called `times_pullback()` while others are called `diag_pullback()`?

Also do we need to be nesting functions here? Could this not just be a variable?

```julia
function rrule(::typeof(*), D::Diagonal{<:Real}, V::AbstractVector{<:Real})
    times_pullback = (NO_FIELDS, @thunk(Diagonal(Ȳ .* V)), @thunk(D.diag .* Ȳ))
    return D.diag * times_pullback
end
```

Reply:
So it would ideally be called `*_pullback`
but that is not a valid function name.
So we subsitute `times` for the `*`

Note: yo u seem to maybe have misread it, we return the function.
we don't multiply by it.
a pullback is always a function, not a value.


The reason it is defared as a named local comment is for if it ever needs debugging.

Constrast the stack-trace
```
julia> function mkanon()
       x = ()->error()
       return x
       end
mkanon (generic function with 1 method)

julia> mkanon()()
ERROR:
Stacktrace:
 [1] error() at ./error.jl:42
 [2] (::var"##7#8")() at ./REPL[1]:2
 [3] top-level scope at REPL[2]:1
```
vs
```
julia> function mkfoo()
       function foo()
           error()
       end
       return foo
       end
mkfoo (generic function with 1 method)

julia> mkfoo()()
ERROR:
Stacktrace:
 [1] error() at ./error.jl:42
 [2] (::var"#foo#9")() at ./REPL[4]:3
 [3] top-level scope at REPL[5]:1
```

The above is on Julia#master,  but evern the LTS does still render the local function names
just uglier.

Reply:
> So it would ideally be called `*_pullback`

We can in principle use the `var` macro/syntax introduced in Julia 1.3 which can be backported easily https://github.com/JuliaLang/julia/pull/32408#issuecomment-506398755:

```julia
julia> VERSION
v"1.0.3"

julia> macro var_str(str)
           Symbol(str)
       end
@var_str (macro with 1 method)

julia> var"*_pullback"() = "hello"
*_pullback (generic function with 1 method)

julia> var"*_pullback"()
"hello"
```

Reply:
Not worth it.